src/manager/node.c

```diff
@@ -1397,105 +1484,154 @@ static void node_send_agent_subscribe_all(Node *node) {
-                                        sub->monitor,
-                                        node->name,
-                                        sub->unit,
-                                        active_state_to_string(usubs->active_state),
-                                        usubs->substate ? usubs->substate : "invalid",
-                                        "virtual");
+                LIST_APPEND(subs, usubs->subs, steal_pointer(&usub));
+
+                /* We know this is loaded, so we won't get notified from
+                   the agent, instead send a virtual event here. */
+                if (usubs->loaded) {
+                        int r = sub->handle_unit_new(sub->monitor, node->name, sub_unit->name, "virtual");
                         if (r < 0) {
                                 hirte_log_error("Failed to emit UnitNew signal");
                         }
+
+                        if (usubs->active_state >= 0) {
+                                r = sub->handle_unit_state_changed(
+                                                sub->monitor,
+                                                node->name,
+                                                sub_unit->name,
+                                                active_state_to_string(usubs->active_state),
+                                                usubs->substate ? usubs->substate : "invalid",
+                                                "virtual");
+                                if (r < 0) {
+                                        hirte_log_error("Failed to emit UnitNew signal");
+                                }
+                        }
                 }
         }
 }
 
-void node_unsubscribe(Node *node, Subscription *sub) {
-        UnitSubscriptionsKey key = { sub->unit };
-        UnitSubscriptions *usubs = NULL;
-        UnitSubscription *usub = NULL;
-        UnitSubscription *found = NULL;
-        UnitSubscriptions *deleted = NULL;
-
-        /* NOTE: If there are errors during subscribe we may still
-           call unsubscribe, so this must silently handle the
-           case of too many unsubscribes. */
-
-        usubs = hashmap_get(node->unit_subscriptions, &key);
-        if (usubs == NULL) {
+void node_subscribe(Node *node, Subscription *sub) {
+        if (subscription_has_unit_wildcard(sub)) {
+                node_subscribe_wildcard_unit(node, sub);
                 return;
         }
+        node_subscribe_units(node, sub);
+}
 
-        LIST_FOREACH(subs, usub, usubs->subs) {
-                if (usub->sub == sub) {
-                        found = usub;
-                        break;
-                }
+void node_unsubscribe_wildcard_unit(Node *node, Subscription *sub) {
+        LIST_REMOVE(wildcard_subscriptions, node->wildcard_subscriptions, sub);
+        if (!node_has_wildcard_subscriptions(node)) {
+                /* First wildcard sub to this node, pass to agent */
```

Comment:
Should comment be "Last wildcard" maybe?

Reply:
Yes, copy & paste error :sweat_smile: